### PR TITLE
Plantuml visualisations

### PR DESCRIPTION
**Pull request by judit-nahaj**
plantuml visualisations
+ updated pdf
[plantuml.pdf](https://github.com/gherking/gherking-docs/files/4841830/plantuml.pdf)

### CODE REVIEW

File: src/data/step.puml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ArrowFontName Segoe UI
+BorderColor black
+FontName Segoe UI
+FontSize 15
+}
+(*) --> "handle Step"
+-right-> "onStep"
+-left-> "handle Step"
+If "argument?"
+--> [Yes] "Process argument"
```
Comment (szikszail) [inline src/data/step.puml]: After processing the step, there is no "process argument" step, it immediately checks the argument type

File: src/data/step.puml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+BackgroundColor #76CDD8
+ArrowColor black
+ArrowFontSize 13
+ArrowFontStyle bold
+ArrowFontName Segoe UI
+BorderColor black
+FontName Segoe UI
+FontSize 15
+}
+(*) --> "handle Step"
```
Comment (szikszail) [inline src/data/step.puml]: I'd add here a label: `-->[step]` to indicate that we got a step as parameter

File: src/data/tags.puml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+@startuml
```
Comment (szikszail) [inline src/data/tags.puml:1]: I think the tag process was a bit different than this, please check it

File: src/data/scenarioOutline.puml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+(*) --> "handle Scenario Outline"
+-right-> "onScenarioOutline"
+-left-> "handle Scenario Outline"
+--> "[*Tags]"
+--> "filter/pre Step"
+-right-> "pre Step"
+-left-> "filter/pre Step"
+--> "Process steps"
+If "Are there unprocessed steps?"
+-left-> [Yes] "Process steps"
```
Comment (szikszail) [inline src/data/scenarioOutline.puml]: I don't think that this "Process steps" part is correct, given that you have the actual processing part in the "No" branch. Same for the "Process examples"

File: src/data/feature.puml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+BackgroundColor #76CDD8
+ArrowColor black
+ArrowFontSize 13
+ArrowFontStyle bold
+ArrowFontName Segoe UI
+BorderColor black
+FontName Segoe UI
+FontSize 15
+}
+(*) --> "Read ?"
```
Comment (szikszail) [inline src/data/feature.puml]: The "Feature" process just receives a `Feature` and processes it, there is no reading in it

File: src/data/feature.puml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+--> "handle tag"
+-right-> "on Tag"
+-left-> "handle tag"
+--> "filter/post Tags"
+-right-> "postfilter Tag"
+-left-> "filter/post Tags"
+--> "Process Elements"
+If "rule?"
+--> [Yes] "applyRule"
+--> "applyElements"
```
Comment (szikszail) [inline src/data/feature.puml]: "Apply Elements" should be in the rule process and not following it

File: src/data/ast.puml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+@startuml
```
Comment (szikszail) [inline src/data/ast.puml]: There is no need for this process

**Review by szikszail — CHANGES_REQUESTED** [review]
Please check my comments. In short:
1. The actual flow is logically not correct in some places
1. I miss the "Rule" process
1. Please use stereotypes to label/color the different types of processes: predefined, outside process (e.g. onStep), etc: https://plantuml.com/activity-diagram-legacy / Skinparams

File: src/data/step.puml (hunk 2 above)
Comment (judit-nahaj) [reply src/data/step.puml]: Should we apply this approach on other files with parameters as well?

Comment (szikszail) [reply src/data/step.puml]: That would be great :)

File: src/data/feature.puml [hunk 8]
```diff
@@ -9,25 +12,35 @@ BorderColor black
...
+-left-> "filter/pre Rule"
+--> "Apply Rule"<<apply>>
+If "Are there rules left?"
+-left->[Yes] "Apply Rule"
 else
---> [No] "applyElements"
+--> [No] "filter/post Rule"
+-right-> "post Rule"
+-left-> "filter/post Rule"<<filter>>
+--> "filter/pre Scenario"<<filter>>
```
Comment (szikszail) [inline src/data/feature.puml]: Handling the rule and the normal elements are mutually exclusive so you either have rules in the feature and you process it, or you have elements (Scenario, Background, etc) and process those, but not the same time.

File: src/data/scenario.puml [hunk 9]
```diff
@@ -9,16 +12,22 @@ BorderColor black
...
+--> "Apply Tags"<<apply>>
+--> "filter/post Tags"<<filter>>
+-right-> "post filter Tags"
+-left-> "filter/post Tags"
 --> "filter/pre Step"
 -right-> "pre Step"
--left-> "filter/pre Step"
---> "apply to Step"
+-left-> "filter/pre Step"<<filter>>
+--> "Apply Step"<<apply>>
```
Comment (szikszail) [inline src/data/scenario.puml:22]: A scenario also has multiple steps.

File: src/data/example.puml [hunk 10]
```diff
@@ -9,20 +12,31 @@ BorderColor black
...
+--> "Apply Tags" <<apply>>
+--> "filter/post Tags" <<filter>>
+-right-> "post Tags"
+-left-> "filter/post Tags"
+--> "filter/pre Row" <<filter>>
 -right-> "pre Row"
--left-> "filter/pre Row"
---> "Process rows"
+-left-> "filter/pre Row" <<filter>>
+--> "Process Rows"
```
Comment (szikszail) [inline src/data/example.puml]: I don't think there is a "process rows" process.

**Review by szikszail — CHANGES_REQUESTED** [review]

**Review by szikszail — APPROVED** [review]

**Pull request merged**